src/domains/TypesDomain.js

```diff
@@ -119,7 +128,7 @@ export default class TypesDomain {
       case "-":
       case "+":
       case "~":
-        resultType = NumberValue;
+        resultType = IntegralValue;
```

Comment:
In the case of - and + you need to know that the operand type is integral before the result type can be declared to be integral. This will require you to change the signature (and the comment) of the method